inference/requirements.txt

```
@@ -1,4 +1,4 @@
-torch==2.4.1
+torch==2.6.0
 triton==3.0.0
```

**Suggested Change to `requirements.txt`**
It seems that `torch 2.6.0`specifically requires `triton 3.2.0` compatibility. Currently, the requirements.txt lists `triton==3.0.0`, which causes a conflict with `torch==2.6.0`.

I recommend updating the `triton` version in the `requirements.txt` file to `3.2.0` to match the requirement from `torch 2.6.0`. The updated section would look like this:

This should resolve the version compatibility issue.

However, please note that `triton==3.2.0` does not currently work on macOs or Windows through `pip`. Users on those platforms will need to manually install Triton by cloning the repository and building it from source. I suggest adding documentation for that in case someone needs to follow those steps. You can follow the installation instructions from the official [Triton repository](https://github.com/triton-lang/triton?tab=readme-ov-file#install-from-source) to do this.

This leads to update `requirement.txt` as follow :
```plaintext
torch==2.6.0
triton==3.2.0; platform_system == "Linux"
transformers==4.46.3
safetensors==0.4.5
```

 🚀

---

Actually further looking into this, Triton does not have a pip package at all for Mac or windows for any version. It currently only supports linux(https://github.com/triton-lang/triton?tab=readme-ov-file#compatibility).

Maybe we can avoid using triton all together and use pure PyTorch but that would come with it's own issues. Not sure how we can proceed.

---

One potential solution is to have `requirements.txt` look something like this:

```suggestion
torch==2.6.0
git+https://github.com/NripeshN/triton.git@main#subdirectory=python
transformers==4.46.3
safetensors==0.4.5
```

This way we install triton from source and should work on every OS, but yea triton would not support GPU acceleration on Macs as of now.

---

While downloading Triton from source using `git+https://github.com/triton-lang/triton@main#subdirectory=python` ensures compatibility across all OS, I am concerned that Triton may still require additional building. 

That said, if we want to fully support Metal on Apple Silicon, we would need to use MLX.
However, MLX is not as mature as Triton and lacks many of the optimizations and utilities that Triton provides for CUDA. Using MLX would require significant custom implementation, particularly for matrix operations and kernel optimizations. We'd need to manually write custom Metal shaders and utilize the Python API to replicate functionality that Triton handles out of the box.

This would involve a substantial amount of work to make MLX work like Triton. It could be a great project, but it'll definitely take some time to get it right.  🚀

---

> While downloading Triton from source using `git+https://github.com/triton-lang/triton@main#subdirectory=python` ensures compatibility across all OS, I am concerned that Triton may still require additional building.
> 
> That said, if we want to fully support Metal on Apple Silicon, we would need to use MLX. However, MLX is not as mature as Triton and lacks many of the optimizations and utilities that Triton provides for CUDA. Using MLX would require significant custom implementation, particularly for matrix operations and kernel optimizations. We'd need to manually write custom Metal shaders and utilize the Python API to replicate functionality that Triton handles out of the box.
> 
> This would involve a substantial amount of work to make MLX work like Triton. It could be a great project, but it'll definitely take some time to get it right. 🚀

Triton is currently built around NVIDIA’s CUDA (NVPTX) and AMD’s ROCm backends. It would not work on other backends unless added(via metal for apple maybe). MLX is an array framework similar to PyTorch. We could implement this entire kernel using PyTorch’s native tensor operations, JIT, or even TorchScript which would enable cross platform, but this could come with some trade-offs like performance etc.